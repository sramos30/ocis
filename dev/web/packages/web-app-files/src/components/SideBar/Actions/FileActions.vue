<template>
  <oc-list id="oc-files-actions-sidebar" class="oc-mt-s">
    <action-menu-item
      v-for="(action, index) in actions"
      :key="`action-${index}`"
      :action="action"
      :items="resources"
      class="oc-py-xs"
    />
  </oc-list>
</template>

<script lang="ts">
import { mapGetters } from 'vuex'
import ActionMenuItem from '../../ActionMenuItem.vue'

import FileActions from '../../../mixins/fileActions'
import { defineComponent } from '@vue/composition-api'

export default defineComponent({
  name: 'FileActions',
  components: { ActionMenuItem },
  mixins: [FileActions],
  computed: {
    ...mapGetters('Files', ['highlightedFile']),

    resources() {
      return [this.highlightedFile]
    },

    actions() {
      return this.$_fileActions_getAllAvailableActions(this.resources)
    }
  }
})
</script>

<style lang="scss">
#oc-files-actions-sidebar {
  > li a,
  > li a:hover {
    color: var(--oc-color-swatch-passive-default);
    display: inline-flex;
    gap: 10px;
    vertical-align: top;
  }
}
</style>
