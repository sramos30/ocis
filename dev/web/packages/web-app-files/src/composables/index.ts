export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './sort'

declare module 'vue/types/vue' {
  interface Vue {
    [key: string]: any
  }
}
