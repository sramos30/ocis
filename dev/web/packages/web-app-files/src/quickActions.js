import { DateTime } from 'luxon'
import copyToClipboard from 'copy-to-clipboard'

export function createPublicLink(ctx) {
  // FIXME: Translate name
  const params = { name: ctx.$gettext('Quick action link'), permissions: 1 }
  const capabilities = ctx.store.state.user.capabilities
  const expirationDate = capabilities.files_sharing.public.expire_date

  if (expirationDate.enabled) {
    params.expireDate = DateTime.now()
      .plus({ days: parseInt(expirationDate.days, 10) })
      .endOf('day')
      .toISO()
  }

  if (ctx.storageId) {
    params.spaceRef = `${ctx.storageId}${ctx.item.path}`
  }

  return new Promise((resolve, reject) => {
    ctx.store
      .dispatch('Files/addLink', {
        path: ctx.item.path,
        client: ctx.client,
        params,
        storageId: ctx.storageId
      })
      .then((link) => {
        ctx.store.dispatch('Files/sidebar/openWithPanel', 'links-item').then(() => {
          copyToClipboard(link.url)
          ctx.store.dispatch('showMessage', {
            title: ctx.$gettext('Public link created'),
            desc: ctx.$gettext(
              'Public link was created successfully and copied into your clipboard.'
            )
          })
        })
        resolve()
      })
      .catch((e) => {
        reject(e)
      })
  })
}

export async function openNewCollaboratorsPanel(ctx) {
  await ctx.store.dispatch('Files/sidebar/openWithPanel', 'sharing-item')
}

export async function openSpaceMembersPanel(ctx) {
  await ctx.store.dispatch('Files/sidebar/openWithPanel', 'space-share-item')
}

export function canShare(item, store) {
  return (
    store.state.user.capabilities.files_sharing &&
    store.state.user.capabilities.files_sharing.api_enabled &&
    item.canShare()
  )
}

export default {
  collaborators: {
    id: 'collaborators',
    label: ($gettext) => $gettext('Add people'),
    icon: 'user-add',
    handler: openNewCollaboratorsPanel,
    displayed: canShare
  },
  publicLink: {
    id: 'public-link',
    label: ($gettext) => $gettext('Create and copy public link'),
    icon: 'link',
    handler: createPublicLink,
    displayed: canShare
  }
}
