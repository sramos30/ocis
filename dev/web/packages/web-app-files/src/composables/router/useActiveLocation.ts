import { ref, Ref, watch } from '@vue/composition-api'
import { useRoute, useRouter } from 'web-pkg/src/composables'
import { ActiveRouteDirectorFunc } from '../../router'

/**
 * watches the current route and re-evaluates the provided active location director
 * on each route name update.
 *
 * @param director
 * @param comparatives
 */
export const useActiveLocation = <T extends string>(
  director: ActiveRouteDirectorFunc<T>,
  ...comparatives: T[]
): Ref<boolean> => {
  const value = ref(false)
  const router = useRouter()
  const currentRoute = useRoute()
  watch(
    currentRoute,
    () => {
      value.value = director(router, ...comparatives)
    },
    { immediate: true }
  )
  return value
}
