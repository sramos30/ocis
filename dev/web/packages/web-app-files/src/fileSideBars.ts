import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import FileShares from './components/SideBar/Shares/FileShares.vue'
import FileLinks from './components/SideBar/Links/FileLinks.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import SpaceDetails from './components/SideBar/Details/SpaceDetails.vue'
import SpaceShares from './components/SideBar/Shares/SpaceShares.vue'
import { isLocationSpacesActive, isLocationTrashActive, isLocationPublicActive } from './router'
import { spaceRoleEditor, spaceRoleManager } from './helpers/share'
import { Panel } from '../../web-pkg/src/components/sidebar'

import { Resource } from './helpers/resource'
import { User } from './helpers/user'
import Router from 'vue-router'

function $gettext(msg: string): string {
  return msg
}

const panelGenerators: (({
  rootFolder,
  highlightedFile,
  router,
  multipleSelection,
  user,
  capabilities
}: {
  rootFolder: boolean
  highlightedFile: Resource
  router: Router
  multipleSelection: boolean
  user: User
  capabilities: any
}) => Panel)[] = [
  // We don't have file details in the trashbin, yet.
  // Only allow `actions` panel on trashbin route for now.
  ({ rootFolder, highlightedFile }): Panel => ({
    app: 'no-selection-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: NoSelection,
    default: () => true,
    get enabled() {
      return rootFolder && highlightedFile?.type !== 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'details-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetails,
    default:
      !isLocationTrashActive(router, 'files-trash-personal') &&
      !isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return (
        !isLocationTrashActive(router, 'files-trash-personal') &&
        !isLocationTrashActive(router, 'files-trash-spaces-project') &&
        !multipleSelection &&
        !rootFolder
      )
    }
  }),
  ({ multipleSelection, rootFolder }) => ({
    app: 'details-multiple-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: FileDetailsMultiple,
    default: () => true,
    get enabled() {
      return multipleSelection && !rootFolder
    }
  }),
  ({ router, highlightedFile }) => ({
    app: 'details-space-item',
    icon: 'questionnaire-line',
    title: $gettext('Details'),
    component: SpaceDetails,
    default: isLocationSpacesActive(router, 'files-spaces-projects'),
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ router, multipleSelection, rootFolder }) => ({
    app: 'actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: FileActions,
    default:
      isLocationTrashActive(router, 'files-trash-personal') ||
      isLocationTrashActive(router, 'files-trash-spaces-project'),
    get enabled() {
      return !multipleSelection && !rootFolder
    }
  }),
  ({ highlightedFile, user }) => ({
    app: 'space-actions-item',
    icon: 'slideshow-3',
    title: $gettext('Actions'),
    component: SpaceActions,
    get enabled() {
      if (highlightedFile?.type !== 'space') {
        return false
      }
      return [
        ...highlightedFile.spaceRoles[spaceRoleManager.name],
        ...highlightedFile.spaceRoles[spaceRoleEditor.name]
      ].includes(user.uuid)
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder }) => ({
    app: 'sharing-item',
    icon: 'group',
    title: $gettext('People'),
    component: FileShares,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.api_enabled
      }
      return false
    }
  }),
  ({ highlightedFile }) => ({
    app: 'space-share-item',
    icon: 'group',
    title: $gettext('Members'),
    component: SpaceShares,
    get enabled() {
      return highlightedFile?.type === 'space'
    }
  }),
  ({ capabilities, router, multipleSelection, rootFolder, highlightedFile }) => ({
    app: 'links-item',
    icon: 'link',
    title: $gettext('Links'),
    component: FileLinks,
    get enabled() {
      if (multipleSelection || (rootFolder && highlightedFile?.type !== 'space')) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }

      if (capabilities.files_sharing) {
        return capabilities.files_sharing.public.enabled
      }
      return false
    }
  }),
  ({ capabilities, highlightedFile, router, multipleSelection, rootFolder }) => ({
    app: 'versions-item',
    icon: 'git-branch',
    title: $gettext('Versions'),
    component: FileVersions,
    get enabled() {
      if (multipleSelection || rootFolder) return false
      if (
        isLocationTrashActive(router, 'files-trash-personal') ||
        isLocationTrashActive(router, 'files-trash-spaces-project') ||
        isLocationPublicActive(router, 'files-public-files')
      ) {
        return false
      }
      return !!capabilities.core && highlightedFile && highlightedFile.type !== 'folder'
    }
  })
]

export default panelGenerators
