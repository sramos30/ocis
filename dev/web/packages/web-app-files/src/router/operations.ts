import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { $gettext, createLocation, isLocationActiveDirector } from './utils'

type operationsTypes =
  | 'files-operations-location-picker'
  | 'files-operations-resolver-private-link'
  | 'files-operations-resolver-public-link'

export const createLocationOperations = (name: operationsTypes, location = {}): Location =>
  createLocation(name, location)

const locationLocationPicker = createLocationOperations('files-operations-location-picker')
const locationResolverPublicLink = createLocationOperations('files-operations-resolver-public-link')
const locationResolverPrivateLink = createLocationOperations(
  'files-operations-resolver-private-link'
)

export const isLocationOperationsActive = isLocationActiveDirector<operationsTypes>(
  locationLocationPicker,
  locationResolverPublicLink,
  locationResolverPrivateLink
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    name: locationLocationPicker.name,
    path: '/ops/location-picker/:context/:action/:item*',
    component: components.LocationPicker,
    meta: {
      verbose: true,
      auth: false,
      patchCleanPath: true
    }
  },
  {
    name: locationResolverPublicLink.name,
    path: '/ops/resolver/public-link/:token',
    component: components.PublicLink,
    meta: {
      auth: false,
      title: $gettext('Resolving public link')
    }
  },
  {
    name: locationResolverPrivateLink.name,
    path: '/ops/resolver/private-link/:fileId',
    component: components.PrivateLink,
    meta: { title: $gettext('Resolving private link') }
  }
]
