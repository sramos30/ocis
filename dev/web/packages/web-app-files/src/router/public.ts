import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes = 'files-public-files' | 'files-public-drop'

export const createLocationPublic = (name: shareTypes, location = {}): Location =>
  createLocation(name, location)

export const locationPublicFiles = createLocationPublic('files-public-files')
export const locationPublicDrop = createLocationPublic('files-public-drop')

export const isLocationPublicActive = isLocationActiveDirector<shareTypes>(
  locationPublicFiles,
  locationPublicDrop
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: '/public',
    component: components.App,
    meta: {
      auth: false
    },
    children: [
      {
        name: locationPublicFiles.name,
        path: 'show/:item*',
        component: components.PublicFiles,
        meta: {
          auth: false,
          title: $gettext('Public files'),
          patchCleanPath: true
        }
      }
    ]
  },
  {
    name: locationPublicDrop.name,
    path: '/public/drop/:token',
    component: components.FilesDrop,
    meta: { auth: false, title: $gettext('Public file upload') }
  }
]
