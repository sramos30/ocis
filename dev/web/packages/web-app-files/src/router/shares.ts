import { RouteComponents } from './router'
import { Location, RouteConfig } from 'vue-router'
import { createLocation, isLocationActiveDirector, $gettext } from './utils'

type shareTypes = 'files-shares-with-me' | 'files-shares-with-others' | 'files-shares-via-link'

export const createLocationShares = (name: shareTypes, location = {}): Location =>
  createLocation(name, location)

export const locationSharesWithMe = createLocationShares('files-shares-with-me')
export const locationSharesWithOthers = createLocationShares('files-shares-with-others')
export const locationSharesViaLink = createLocationShares('files-shares-via-link')

export const isLocationSharesActive = isLocationActiveDirector<shareTypes>(
  locationSharesWithMe,
  locationSharesWithOthers,
  locationSharesViaLink
)

export const buildRoutes = (components: RouteComponents): RouteConfig[] => [
  {
    path: '/shares',
    component: components.App,
    redirect: locationSharesWithMe,
    children: [
      {
        name: locationSharesWithMe.name,
        path: 'with-me',
        component: components.SharedWithMe,
        meta: {
          title: $gettext('Files shared with me')
        }
      },
      {
        name: locationSharesWithOthers.name,
        path: 'with-others',
        component: components.SharedWithOthers,
        meta: {
          title: $gettext('Files shared with others')
        }
      },
      {
        name: locationSharesViaLink.name,
        path: 'via-link',
        component: components.SharedViaLink,
        meta: {
          title: $gettext('Files shared via link')
        }
      }
    ]
  }
]
