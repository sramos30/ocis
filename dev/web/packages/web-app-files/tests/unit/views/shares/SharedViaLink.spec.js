import { shallowMount, mount } from '@vue/test-utils'
import { getStore, localVue, createFile } from '../views.setup.js'
import { createLocationSpaces } from '../../../../src/router'
import FileActions from '@files/src/mixins/fileActions'
import SharedViaLink from '@files/src/views/shares/SharedViaLink.vue'

const component = { ...SharedViaLink, mounted: jest.fn() }

const router = {
  push: jest.fn(),
  afterEach: jest.fn(),
  currentRoute: {
    name: 'some-route-name',
    query: {}
  },
  resolve: (r) => {
    return { href: r.name }
  }
}

const resources = [
  createFile({ id: 2147491323, type: 'file' }),
  createFile({ id: 2147491324, type: 'file' })
]

const stubs = {
  'app-bar': true,
  'resource-table': false,
  'context-actions': true,
  pagination: true,
  'list-info': true,
  'router-link': true
}

const listLoaderStub = 'app-loading-spinner-stub'
const listInfoStub = 'list-info-stub'
const contextActionsStub = 'context-actions-stub'

const selectors = {
  noContentMessage: '#files-shared-via-link-empty',
  ocTableFiles: '#files-shared-via-link-table'
}

describe('SharedViaLink view', () => {
  const spyTriggerDefaultAction = jest
    .spyOn(FileActions.methods, '$_fileActions_triggerDefaultAction')
    .mockImplementation()
  const spyRowMounted = jest.spyOn(SharedViaLink.methods, 'rowMounted')

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('when the view is still loading', () => {
    it('should show app-loading-spinner component', () => {
      const wrapper = getShallowWrapper({ loading: true })
      const listLoader = wrapper.find(listLoaderStub)

      expect(listLoader.exists()).toBeTruthy()
    })
  })

  describe('when the view is not loading anymore', () => {
    it('should not app-loading-spinner component', () => {
      const wrapper = getShallowWrapper()
      expect(wrapper.find(listLoaderStub).exists()).toBeFalsy()
    })

    describe('when there are no files to be displayed', () => {
      let wrapper
      beforeEach(() => {
        wrapper = getMountedWrapper({ stubs })
      })

      it('should show no-content-message component', () => {
        const noContentMessage = wrapper.find(selectors.noContentMessage)

        expect(noContentMessage.exists()).toBeTruthy()
        expect(wrapper).toMatchSnapshot()
      })

      it('should not show oc-table-files component', () => {
        expect(wrapper.find(selectors.ocTableFiles).exists()).toBeFalsy()
      })
    })

    describe('when there are one or more files to be displayed', () => {
      let wrapper
      beforeEach(() => {
        const store = createStore({
          totalFilesCount: { files: resources.length, folders: 0 }
        })
        wrapper = getMountedWrapper({
          store,
          setup: {
            paginatedResources: resources
          }
        })
      })

      it('should not show no-content-message component', () => {
        expect(wrapper.find(selectors.noContentMessage).exists()).toBeFalsy()
      })
      it('should show oc-table-files component with props', () => {
        const ocTableFiles = wrapper.find(selectors.ocTableFiles)

        expect(ocTableFiles.exists()).toBeTruthy()
        expect(ocTableFiles.props().resources).toMatchObject(resources)
        expect(ocTableFiles.props().areThumbnailsDisplayed).toBe(false)
        expect(ocTableFiles.props().headerPosition).toBe(0)
        expect(ocTableFiles.props().targetRoute).toMatchObject(
          createLocationSpaces('files-spaces-personal-home')
        )
      })
      it('should set props on list-info component', () => {
        const listInfo = wrapper.find(listInfoStub)

        expect(listInfo.props().files).toEqual(resources.length)
        expect(listInfo.props().folders).toEqual(0)
      })
      it('should trigger the default action when a "fileClick" event gets emitted', () => {
        const ocTableFiles = wrapper.find(selectors.ocTableFiles)

        expect(spyTriggerDefaultAction).toHaveBeenCalledTimes(0)

        ocTableFiles.vm.$emit('fileClick')

        expect(spyTriggerDefaultAction).toHaveBeenCalledTimes(1)
      })
      it('should lazily load previews when a "rowMounted" event gets emitted', () => {
        expect(spyRowMounted).toHaveBeenCalledTimes(resources.length)
      })
      it('should not show context actions', () => {
        const contextActions = wrapper.find(contextActionsStub)

        expect(contextActions.exists()).toBeFalsy()
      })

      describe('when a file is highlighted', () => {
        it('should set props on context-actions component', () => {
          const selectedFiles = [resources[0]]
          const store = createStore({
            totalFilesCount: { files: resources.length, folders: 0 },
            selectedFiles: selectedFiles
          })
          const wrapper = getMountedWrapper({
            store: store,
            setup: {
              paginatedResources: resources
            }
          })
          const contextActions = wrapper.find(contextActionsStub)

          expect(contextActions.exists()).toBeTruthy()
          expect(contextActions.props().items).toMatchObject(selectedFiles)
        })
      })
    })
  })
})

function mountOptions(store, loading, setup = {}) {
  return {
    localVue,
    store: store,
    stubs,
    mocks: {
      $route: router.currentRoute,
      $router: router
    },
    setup: () => ({
      loadResourcesTask: {
        isRunning: loading,
        perform: jest.fn()
      },
      ...setup
    })
  }
}

function getMountedWrapper({ store = createStore(), loading = false, setup } = {}) {
  return mount(component, mountOptions(store, loading, setup))
}

function getShallowWrapper({ store = createStore(), loading = false, setup } = {}) {
  return shallowMount(component, mountOptions(store, loading, setup))
}

function createStore({ totalFilesCount, highlightedFile, selectedFiles } = {}) {
  return getStore({
    highlightedFile,
    totalFilesCount,
    selectedFiles
  })
}
