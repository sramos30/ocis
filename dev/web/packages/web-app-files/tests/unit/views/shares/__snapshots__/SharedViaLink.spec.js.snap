// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedViaLink view when the view is not loading anymore when there are no files to be displayed should show no-content-message component 1`] = `
<div>
  <app-bar-stub breadcrumbs="" breadcrumbscontextactionsitems="" hassharesnavigation="true" hassidebartoggle="true" hasviewoptions="true"></app-bar-stub>
  <div class="oc-height-1-1 oc-flex oc-flex-column oc-flex-center oc-flex-middle oc-text-center files-empty" id="files-shared-via-link-empty">
    <div class="oc-icon oc-icon-xxl oc-icon-passive">
      <!---->
    </div>
    <div class="oc-text-muted oc-text-large"><span>There are no resources with a public link at the moment</span></div>
    <div class="oc-text-muted"></div>
  </div>
</div>
`;
